### ISSUE TITLE
Segmentation fault with RedisCluster::rawCommand with BLPOP, XREAD, etc

### ISSUE THREAD

**Issue by offshore** [phpredis/phpredis#1847]
The title should be descriptive enough.

Assuming `$key` does not exist:
```
$redis->blPop($key, 1); // returns []
$redis->rawCommand("BLPOP", $key, 1); // returns []
$redisCluster->blPop($key, 1); // returns []
$redisCluster->blPop($key, 1); // segmentation fault
```

This behavior can be seen on any version and environment due to incorrect handling of redis protocol's null multibulk replies by PHPRedis: it affects any command that is replied by redis server with `"*-1\r\n"` (e.g. blocking read commands with empty result: `BLPOP`, `XREAD`, `XREADGROUP`, etc). The bug does also lead to incorrect return value from such commands: it should be `null`, not an empty array.

Pull request with test case and patch that fixes protocol handling for `RedisCluster::rawCommand` case will be opened shortly. The wrapping functions (`blPop`, `xRead`, ...) should be fixed separately, as well as a regular client (_maybe_: I did no deep investigation for that). I think, the documentation should be fixed too. I did nothing more than a basic things to fix up a mess: this can break stuff which relies on `[]` instead of `null` in return values and I'm not quite familiar to your code and workflow.

**Comment by offshore** [comment]
FYI: https://redis.io/topics/protocol#resp-arrays

**Comment by michael-grunder** [comment]
Thanks for the PR.

I am aware of the Redis protocol specification as we handle null multi bulk responses all over phpredis (and in hiredis, which I also maintain).

It is surprising that no one has encountered this bug before though.  Using `RedisCluster` + `rawCommand()` + `MULTI..EXEC` must be quite rare otherwise we'd have seen this a long time ago.  The bug has been here for at least 5 years.

**Comment by offshore** [comment]
Any chance that it would be finally merged in near future? Anything more that I can help with?

**Comment by yatsukhnenko** [comment]
I think we will fix this before next release 🙂

**Comment by michael-grunder** [comment]
The code looks good.

I'm going to get it merged but I think we need to return NULL MULTIBULK replies as empty arrays to stay consistent with the rest of PhpRedis.

NULL Is correct per the Redis specification but I think users of the library will expect empty arrays (because that's how it's handled in other methods).  We should fix this in PhpRedis 6 though as we can make breaking changes.

@yatsukhnenko What do you think?

Edit:  Another option which would be a bit more work but not too bad is to add `Redis::OPT_NULL_MULTIBULK_AS_NULL` that can be configured via `setOption()` similar to the way we let users request literal status strings.  This would allow both the correct behavior per the redis spec without breaking changes.

**Comment by yatsukhnenko** [comment]
@michael-grunder I like the idea of using `Redis::OPT_NULL_MULTIBULK_AS_NULL` option.

**Comment by michael-grunder** [comment]
OK, I'll add that to the PR and get it implemented this weekend.

**Comment by yatsukhnenko** [comment]
@michael-grunder about changes in PR, maybe we should change type of [elements](https://github.com/phpredis/phpredis/blob/develop/cluster_library.h#L328) from `size_t` to `ssize_t` to allow store `-1` as number of elements

**Comment by michael-grunder** [comment]
That also crossed my mind.

**Comment by michael-grunder** [comment]
I've got the changes implemented locally with the addition of an option that tells phpredis to return null multibulk replies as NULL.

I'll have the complete fix up tomorrow

**Comment by michael-grunder** [comment]
[Branch here](https://github.com/phpredis/phpredis/commits/issue.1847-cluster-segfault)

**Comment by michael-grunder** [comment]
Closed via #1850 

Feel free to reopen the issue if you encounter issues with the fix.
